src/main/java/frc/robot/subsystems/DriveBaseSubsystem.java

```diff
@@ -61,11 +65,57 @@ public DriveBaseSubsystem(Joystick joystick) {
     // differential drive
     m_differentialDrive = new DifferentialDrive(m_motorControllers[Constants.driveLeftFrontIndex].getSparkMax(), 
                                           m_motorControllers[Constants.driveRightFrontIndex].getSparkMax());
+
+    this.usingExternal = usingExternal; // gets key if using external or internal encoders
+
+    // external encoders            
+    m_extLeftEncoder = new Encoder(Constants.leftEncoderDIOone, Constants.leftEncoderDIOtwo, 
+    false, Encoder.EncodingType.k2X); // TODO: confirm correct configuration for encoder
+    m_extRightEncoder = new Encoder(Constants.rightEncoderDIOone, Constants.rightEncoderDIOtwo, 
+    false, Encoder.EncodingType.k2X);
+
+    m_extLeftEncoder.setReverseDirection(true);
+    
+    // internal encoders
+    m_internalLeftEncoder = m_motorControllers[Constants.driveLeftFrontIndex].getEncoder();
+    m_internalRightEncoder = m_motorControllers[Constants.driveRightFrontIndex].getEncoder();
+
+    m_internalLeftEncoder.setPositionConversionFactor(2 * Math.PI * Constants.wheelRadius / Constants.inchesInMeter); // calculate circumference then convert to meters
+    m_internalRightEncoder.setPositionConversionFactor(2 * Math.PI * Constants.wheelRadius / Constants.inchesInMeter);
+
+    // tuning PID, experimental values, multiply by 2
+    SmartDashboard.putNumber(m_motorControllers[Constants.driveLeftFrontIndex].getName() + " P Value", 0.000005);
+    SmartDashboard.putNumber(m_motorControllers[Constants.driveLeftRearIndex].getName() + " P Value", 0.000005);
+    SmartDashboard.putNumber(m_motorControllers[Constants.driveRightFrontIndex].getName() + " P Value", 0.000005);
+    SmartDashboard.putNumber(m_motorControllers[Constants.driveRightRearIndex].getName() + " P Value", 0.000005);
+
+
+    SmartDashboard.putNumber(m_motorControllers[Constants.driveRightRearIndex].getName() + " I Value", 0.0000025);
+    SmartDashboard.putNumber(m_motorControllers[Constants.driveRightRearIndex].getName() + " I Value", 0.0000025);
+    SmartDashboard.putNumber(m_motorControllers[Constants.driveRightRearIndex].getName() + " I Value", 0.0000025);
+    SmartDashboard.putNumber(m_motorControllers[Constants.driveRightRearIndex].getName() + " I Value", 0.0000025);
```

Comment:
A. You should only need to set the leader motor on each side 
B. Since you have an array of motors, this is better done with a for-loop